content/blog/2024-08-29-valkey-memory-efficiency-8-0.md

```diff
@@ -0,0 +1,148 @@
+* [Random key for eviction](https://valkey.io/topics/lru-cache/) - The server does random sampling of the keyspace to find ideal candidate for eviction.
+* [Finding a random key](https://valkey.io/commands/randomkey/) - Command like `RANDOMKEY` retrieves a random key from the database.
+
+
+It is quite possible that some slots contains more keys than others and some could be also empty which could lead the engine to traverse through the all the empty dictionary for a full scan i.e. [O(N)](https://en.wikipedia.org/wiki/Big_O_notation) operation if approached naively. When picking a random key for eviction or running `RANDOMKEY` operation, the server needs to pick a non-empty slot in a performant manner. These cases require a data structure which provides the following functionality:
+
+1. Modify value for a given index (slot) - If a key gets added or removed, increment or decrement the value for that given slot by 1 respectively.
+2. Cumulative frequency until each index (slot) - For a given number representing a key between 1 and the total number of keys, return the index (slot) which covers the particular key.
+
+If approached naively, the former and latter operation would take O(1) and O(N) respectively. However, we want to minimize the latter operation’s time complexity and avoid regression in the operations stated above. Hence, a [binary indexed tree (BIT) or fenwick tree](https://www.topcoder.com/thrive/articles/Binary%20Indexed%20Trees) which provides the above functionality with a minimal memory overhead (~1 MB per node) and the time complexity is also bounded to O(M log N) for both the operations where M = number of modification(s) and N = number of slots. This enables skipping over empty slot efficiently while iterating over the keyspace as well as finding a slot for a given key index in logarithmic time via binary search over the cummulative sum maintained by the BIT. 
+
+Overall with this new approach, the benefits are: 
+
+1. Removes additional memory overhead in cluster mode: Get rid of two pointers (16 bytes) per key to keep the mapping of slot to keys.
+2. Rehashing operation is CPU intensive: By default, limited number of buckets are allocated in a dictionary and it expands/shrinks dyanmically based on usage. While undergoing rehashing, all the data needs to be moved from an old dictionary to a new dictionary. With Valkey 7.2, a global dictionary being shared across all the slots, all the keys gets stored under a single dictionary and each time the fill factor (number of keys / number of buckets) goes above 1.5, the dictionary needs to a large dictionary (multiple of 2) and move large amount of keys. As this operation is performed on the fly, it causes increase in latency for regular command operations while it’s ongoing. With this optimization, the impact of rehashing is localized to the specific dictionary undergoing the process and only subset of key needs to be moved. 
```

Comment:
The verb is missing here: 
`the dictionary needs to a large dictionary` -> `the dictionary needs to move to a larger dictionary` (I think...)